cli/src/commands/auth.rs

```diff
@@ -0,0 +1,50 @@
+use crate::commands::Command;
+use crate::config::CliConfig;
+use crate::errors::{ExitCode, Fallible};
+use crate::style::KEY;
+use crate::telemetry::Session;
+use crate::terminal::sensitive;
+use console::style;
+use log::{debug, error, info, warn};
+use structopt::StructOpt;
+
+#[derive(StructOpt)]
+#[structopt(rename_all = "kebab-case")]
+pub enum Auth {
+    /// Link the CLI to your Apollo account using your Personal API Key.
+    Setup(Setup),
+}
+
+#[derive(StructOpt)]
+pub struct Setup {}
+
+impl Command for Setup {
+    fn run(&self, session: &mut Session) -> Fallible<ExitCode> {
+        session.log_command("auth setup");
+
+        debug!("Checking is config's api_key is already set...");
+        if session.config.api_key.is_some() {
+            warn!("Authentication already configured.");
+        }
+
+        info!("To link your CLI to your Apollo account go to {} and create a new Personal API Key. Once you've done that, copy the key and paste it into the prompt below.",
+            style("https://engine.apollographql.com/user-settings").cyan());
+        let key = sensitive("User key:")?;
+
+        debug!("Checking user input...");
+        if key.is_empty() {
+            error!("No key was inputed, quitting without changes.");
+            return Ok(ExitCode::ConfigurationError);
+        }
+
+        debug!("Setting new key...");
+        let mut config = session.config.clone();
+        config.api_key = Some(key);
```

Comment:
So we thought about doing that but also felt weird/gross doing that as we'd be mutating state that potentially others are relying upon. Maybe a better way to communicate that is by making the `session`'s `config` immutable?

Mind you, the call stack here should almost always be painfully short, but I think there's value in making the side-effecty stuff limited.

Reply:
Unfortunately (or fortunately, depends how you look at it :)) Rust doesn't have an easy way to define mutability at the field level.

> mutating state that potentially others are relying upon

Can you elaborate here? What use case did you have in mind?
I'm as new to this language as you, but it's my understanding that Rust has compiler level support to ensure no 2 pieces of code have access to the same mutable instance at the same time, and having (aliases to the same underlying data is very bad, and not easy to code)

Reply:
1. `derive` for `clone` is much cleaner, many thanks for the tidbit!
2. That sounds good. My original reason for reaching for the hard `clone` was that it signals that you _shouldn't_ do this and as a way to avoid side effects (ie, for those new to things, it's potentially clearer). I also didn't know that about the compiler. I'm game for refactoring this and we can always go back as it's just bits and bytes.